[backends] Add options for the param --category in help message for e…

harshalmittal4 opened the pull request:
Addresses #511 

This PR adds the different available options for the param `--category` in the help message for each backend.

inishchith commented:
@harshalmittal4 This looks to me as a nice addition. 
I was thinking of something like we could get for instance [CATEGORY](https://github.com/chaoss/grimoirelab-perceval/blob/70f82e64badf04e7de63b402107ba211e3be2ced/perceval/backends/core/github.py#L90) variable and update the `helptext` such as.

```python3
helptext = "type of the items to fetch ({categories})".format(categories=", ".join(GitHub.CATEGORIES))
```

This would avoid updating the `helptext` every time there's some change ( or update in categories of data retrieved ) in implementation in future.
Just a suggestion :)

@valeriocos what do you think?

valeriocos commented:
Thank you @harshalmittal4 for the PR, the idea is good but the PR requires some refinements. I agree with the suggestion of @inishchith , which seems easier to maintain.

A possibility could be to change `setup_cmd_parser` from staticmethod to classmethod, and pass the categories value in this way:

```
class AskbotCommand(BackendCommand):
    """Class to run Askbot backend from the command line."""

    BACKEND = Askbot

    @classmethod
    def setup_cmd_parser(cls):
        """Returns the Askbot argument parser."""

        parser = BackendCommandArgumentParser(from_date=True,
                                              archive=True,
                                              categories=cls.BACKEND.CATEGORIES)

        # Required arguments
        parser.parser.add_argument('url',
                                   help="URL of the Askbot server")

        return parser
```

Then the `BackendCommandArgumentParser` should be changed to something like:
```
    def __init__(self, from_date=False, to_date=False, offset=False,
                 basic_auth=False, token_auth=False, archive=False,
                 aliases=None, categories=None):
        self._from_date = from_date
        self._to_date = to_date
        self._archive = archive

        self.aliases = aliases or {}
        self.parser = argparse.ArgumentParser()

        group = self.parser.add_argument_group('general arguments')
        group.add_argument('--category', dest='category',
                           help="type of the items to fetch (%s)" % ','.join(categories))
```

what do you think @sduenas ?

harshalmittal4 commented:
Actually, I was working on refining the PR(by modifying the tests) and trying to add some tests for the proposed change. 
@valeriocos, @inishchith, the idea is better and more structured definitely. But the thing that remains in this is that the param `category` is still be shown under the heading `general arguments` instead of `<backend> arguments` in the helptext. How can we correct this, if we go this way.

harshalmittal4 commented:
And yes, thanks @inishchith, really good having your suggestions :sweat_smile:

harshalmittal4 commented:
@valeriocos ping :)

valeriocos commented:
@harshalmittal4 , I guess there is no problem in having the category param in the general arguments (since it's an argument valid for all backends). The backend params should be those ones that are backend-specific.

What do you think?

harshalmittal4 commented:
@valeriocos, as we are making the helptext for `--category` to be backend-specific (as it will include the different options that can be passed for param `--category` in that backend), so I initially thought to try including it in the group <backend> arguments.

harshalmittal4 commented:
@valeriocos, as we are making the helptext for --category to be backend-specific (as it will include the different options that can be passed for param --category in that backend), so I initially thought to try including it in the group backend arguments.

harshalmittal4 commented:
@valeriocos, as we are making the helptext for --category to be backend-specific (as it will include the different options that can be passed for param --category in that backend), so I initially felt iincluding it in the group backend arguments would be better :slightly_smiling_face:
But the above suggested way seems better with regards to implementation.
So shall I modify the PR as suggested ?

valeriocos commented:
Yes please @harshalmittal4 . 
@jgbarah @sduenas any feedback on this PR?

harshalmittal4 commented:
@valeriocos, I modified the PR as required, 
@jgbarah, @sduenas, please see to this, thanks :)

valeriocos commented on perceval/backends/core/askbot.py:
```diff
@@ -526,12 +526,12 @@ class AskbotCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
please add a param per line, for instance:
```
parser = BackendCommandArgumentParser(from_date=True,
                                      archive=True, 
                                      categories=cls.BACKEND.CATEGORIES)
```

valeriocos commented on perceval/backends/core/bugzilla.py:
```diff
@@ -347,13 +347,13 @@ class BugzillaCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               basic_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above, please add a param per line

valeriocos commented on perceval/backends/core/bugzillarest.py:
```diff
@@ -453,14 +453,14 @@ class BugzillaRESTCommand(BackendCommand):
                                               basic_auth=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/confluence.py:
```diff
@@ -282,12 +282,12 @@ class ConfluenceCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/discourse.py:
```diff
@@ -399,13 +399,13 @@ class DiscourseCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/dockerhub.py:
```diff
@@ -205,11 +205,11 @@ class DockerHubCommand(BackendCommand):
         """Returns the DockerHub argument parser."""
 
-        parser = BackendCommandArgumentParser(archive=True)
+        parser = BackendCommandArgumentParser(archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/gerrit.py:
```diff
@@ -491,12 +491,12 @@ class GerritCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/git.py:
```diff
@@ -341,11 +341,11 @@ def _pre_init(self):
         """Returns the Git argument parser."""
 
-        parser = BackendCommandArgumentParser(from_date=True, to_date=True)
+        parser = BackendCommandArgumentParser(from_date=True, to_date=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/github.py:
```diff
@@ -877,14 +877,14 @@ class GitHubCommand(BackendCommand):
                                               to_date=True,
                                               token_auth=False,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/gitlab.py:
```diff
@@ -628,13 +628,13 @@ class GitLabCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/googlehits.py:
```diff
@@ -238,11 +238,11 @@ class GoogleHitsCommand(BackendCommand):
         """Returns the GoogleHits argument parser."""
 
-        parser = BackendCommandArgumentParser(archive=True)
+        parser = BackendCommandArgumentParser(archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/groupsio.py:
```diff
@@ -267,12 +267,12 @@ def _pre_init(self):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              token_auth=True)
+                                              token_auth=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/hyperkitty.py:
```diff
@@ -264,11 +264,11 @@ def _pre_init(self):
         """Returns the HyperKitty argument parser."""
 
-        parser = BackendCommandArgumentParser(from_date=True)
+        parser = BackendCommandArgumentParser(from_date=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/jenkins.py:
```diff
@@ -246,11 +246,11 @@ class JenkinsCommand(BackendCommand):
         """Returns the Jenkins argument parser."""
 
-        parser = BackendCommandArgumentParser(archive=True)
+        parser = BackendCommandArgumentParser(archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/jira.py:
```diff
@@ -409,13 +409,13 @@ class JiraCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               basic_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/launchpad.py:
```diff
@@ -451,13 +451,13 @@ class LaunchpadCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               archive=True,
-                                              token_auth=False)
+                                              token_auth=False, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/mattermost.py:
```diff
@@ -385,13 +385,13 @@ class MattermostCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/mbox.py:
```diff
@@ -328,11 +328,11 @@ class MBoxCommand(BackendCommand):
         """Returns the MBox argument parser."""
 
-        parser = BackendCommandArgumentParser(from_date=True)
+        parser = BackendCommandArgumentParser(from_date=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/mediawiki.py:
```diff
@@ -528,12 +528,12 @@ class MediaWikiCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/meetup.py:
```diff
@@ -270,14 +270,14 @@ class MeetupCommand(BackendCommand):
                                               to_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/nntp.py:
```diff
@@ -361,12 +361,12 @@ class NNTPCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(offset=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/phabricator.py:
```diff
@@ -612,13 +612,13 @@ class PhabricatorCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/pipermail.py:
```diff
@@ -144,11 +144,11 @@ def _pre_init(self):
         """Returns the Pipermail argument parser."""
 
-        parser = BackendCommandArgumentParser(from_date=True)
+        parser = BackendCommandArgumentParser(from_date=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/redmine.py:
```diff
@@ -287,13 +287,13 @@ class RedmineCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/rss.py:
```diff
@@ -176,11 +176,11 @@ class RSSCommand(BackendCommand):
         """Returns the RSS argument parser."""
 
-        parser = BackendCommandArgumentParser(archive=True)
+        parser = BackendCommandArgumentParser(archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/slack.py:
```diff
@@ -431,13 +431,13 @@ class SlackCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/stackexchange.py:
```diff
@@ -301,13 +301,13 @@ class StackExchangeCommand(BackendCommand):
         parser = BackendCommandArgumentParser(from_date=True,
                                               token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/supybot.py:
```diff
@@ -264,15 +264,15 @@ class SupybotCommand(BackendCommand):
         }
         parser = BackendCommandArgumentParser(from_date=True,
-                                              aliases=aliases)
+                                              aliases=aliases, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/telegram.py:
```diff
@@ -269,7 +269,7 @@ def setup_cmd_parser():
                                               token_auth=True,
                                               archive=True,
-                                              aliases=aliases)
+                                              aliases=aliases, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on perceval/backends/core/twitter.py:
```diff
@@ -343,12 +343,12 @@ class TwitterCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(token_auth=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
same comment as above

valeriocos commented on testing.json:
```diff
@@ -0,0 +1,9 @@
+{"backend_name":"GitHub","backend_version":"0.20.1","category":"pull_request","classified_fields_filtered":null,"data":{"_links":{"comments":{"href":"https://api.github.com/repos/Cloud-CV/EvalAI/issues/1/comments"},"commits":{"href":"https://api.github.com/repos/Cloud-CV/EvalAI/pulls/1/commits"},"html":{"href":"https://github.com/Cloud-CV/EvalAI/pull/1"},"issue":{"href":"https://api.github.com/re ...
```
why this file is needed?

harshalmittal4 replied:
Sorry, it was for some other checking, by  mistake.

harshalmittal4 replied:
Sorry my mistake @valeriocos, it was for some other checking, removed.

valeriocos commented on perceval/backend.py:
```diff
@@ -291,7 +291,7 @@ class BackendCommandArgumentParser:
+                 aliases=None, categories=[]):
         self._from_date = from_date
         self._to_date = to_date
         self._archive = archive
```
I would store the categories in a variable to simplify the tests, something like:
`self._categories = categories`

what do you think ?

harshalmittal4 replied:
Agreed @valeriocos, its more organised, I will update, thanks :)

valeriocos commented:
please, @jgbarah @sduenas provide feedback on this PR when you have time.

sduenas commented on perceval/backends/core/askbot.py:
```diff
@@ -526,12 +526,12 @@ class AskbotCommand(BackendCommand):
 
         parser = BackendCommandArgumentParser(from_date=True,
-                                              archive=True)
+                                              archive=True, categories=cls.BACKEND.CATEGORIES)
```
I would put `categories` in a new line to align it with the style followed for the rest of the parameters. Moreover, if we change the definition of `setup_cmd_parser`, it should be also modified in the parent class.

harshalmittal4 replied:
Done @sduenas!

harshalmittal4 commented on perceval/backend.py:
```diff
@@ -291,7 +291,7 @@ class BackendCommandArgumentParser:
     def __init__(self, from_date=False, to_date=False, offset=False,
                  basic_auth=False, token_auth=False, archive=False,
-                 aliases=None):
+                 aliases=None, categories=[]):
```
Done, please review.

harshalmittal4 commented:
Hey @sduenas @valeriocos
I have updated and added the tests assuming `categories` to be a required argument, want some guidance on this PR for changes required if any, please review..
(The commits squashed and rebased with master)

valeriocos commented on tests/test_backend.py:
```diff
@@ -539,15 +541,21 @@ class TestBackendCommandArgumentParser(unittest.TestCase):
         self.assertIsInstance(parser.parser, argparse.ArgumentParser)
 
+    def test_categories(self):
+        """Test whether _categories is initialised"""
```
Please, add an empty line between the comment and the code. In addition, I guess initialised should be initialized (a quick search on the perceval repo shows that `initialized` is the preferred option, https://github.com/chaoss/grimoirelab-perceval/search?q=initialized&unscoped_q=initialized)

harshalmittal4 replied:
Done, thanks :)

harshalmittal4 commented:
@valeriocos, @sduenas please review when you have time, thanks :)

valeriocos commented:
Thanks @harshalmittal4 , LGTM. What do you think @sduenas  ?

harshalmittal4 commented:
I also felt this earlier once, @sduenas, updated :)
